### PR TITLE
ADD aria attributes for better accessibility...

### LINKED ISSUE #1183: Using JW Player with VoiceOver is impossible
### Expected Behavior

We would like to be able to use JW Player with iOS and Mac OS' VoiceOver...
The different buttons of the player should be labelled (and translated too) for the screen reader softwares.

### Actual Behavior

The buttons of the player are empty and only contain an icon injected via CSS

### User story

We received a complaint from a blind visitor.

He is using Apple's VoiceOver screen reader in order to surf to our sites on his Mac and on his iPhone and he is having difficulties to play the videos.

We are using  on our site auvio:
http://www.rtbf.be/auvio/detail_?id=2106569

Or even easier, you can simply open the src of the iframe containing the video player with this url
http://www.rtbf.be/auvio/embed/media?id=2106569&autoplay=0

When I try it by myself using my iPhone 6S with VoiceOver, sliding my finger blindly accross the screen, I don't get much useful VoiceOver description when I'm touching the video...

I hear "Afficher les contrôles, bouton" ("Show the controls, button") when I'm near the bottom area of the video area...
The big play button is generated with this markup:

`<div class="jw-icon jw-icon-display jw-button-color jw-reset"></div>`

This markup does not contains any text consummable by the screen readers such as the Twitter Bootstrap utility class like

`<span class="sr-only">Play the video</span>`

or a similar aria attribute...

What can we do in order to improve the VoiceOver experience knowing that the markup is generated by the JW Player js script ?

Thank you very much

### Environment

- MacBook Pro and iPhone 6S
- Mac OS Yosemite and iOS 9
- JWPlayer v7.3.6 with ads

----------------------------------------------------------------------------------------

This template is provided as a guide to filing bugs and feature requests. Please replace 
all text *in italics* with details describing your issue. When submitting a feature request, 
"Steps to reproduce" can be substituted with a user story or list of acceptance criteria.

### PR DESCRIPTION
**Pull request by francoismassart**
### Changes proposed in this pull request:

Make the JS version of jwplayer compatible with Voice Over via `aria-label` attributes

Fixes #1183 
JW7-2628

-Labels can be translated via the config
-aria-label
-aria-hidden (when needed)
-role
-tabindex

### CODE REVIEW

**Comment by robwalch** [comment]
Thanks @francoismassart! I'm reviewing with the team and determining which release this can make it in to.

**Comment by francoismassart** [comment]
No problem if you edit my changes but one thing that matter to us is to be able to translate the `aria-label`'s. Our sites are in French, so are our blind users...
Thank you for the feedback.

**Comment by robwalch** [comment]
Yes. We would have liked to figure out internationalization for the player before the labels but hopefully this will get things going in that direction.

I'd like to make the setup option multi-purpose and use simpler names that don't match the CSS classes and allow us to use dot syntax. So for config `Defaults` it would look more like:
```js
localization: {
  play: 'Play',
  playback: 'Start playback',
  pause: 'Pause',
  volume: 'Volume',
  prev: 'Previous',
  next: 'Next',
  cast: 'Chromecast',
  fullscreen: 'Fullscreen',
  playlist: 'Playlist',
  hd: 'Quality',
  cc: 'Closed captions',
  audioTracks: 'Audio tracks', // need to convert camelCase to hyphens
  replay: 'Replay',
  buffer: 'Loading'
}
```

This should still allow you to provide localization strings for the aria tags en Français ;)

**Comment by francoismassart** [comment]
Please add also the "live status" translation which is currently harcoded to "Live broadcast" ;-)
Or I can do another PR once you merged the aria labels.

**Comment by robwalch** [comment]
Yes. I'll inline my comment from above.

File: src/js/api/config.js [hunk 1]
```diff
@@ -19,7 +19,23 @@ define([
...
+            'jw-icon-next': 'Next',
+            'jw-icon-cast': 'Chromecast',
+            'jw-icon-fullscreen': 'Fullscreen',
+            'jw-icon-playlist': 'Playlist',
+            'jw-icon-hd': 'Quality',
+            'jw-icon-cc': 'Closed captions',
+            'jw-icon-audio-tracks': 'Audio tracks',
+            'jw-icon-replay': 'Replay',
+            'jw-icon-buffer': 'Loading'
+        }
```
Comment (robwalch) [inline src/js/api/config.js]: Let's use simpler names that don't match the CSS classes and allow for dot syntax. This `aria` block should look like:
```js
localization: {
  play: 'Play',
  playback: 'Start playback',
  pause: 'Pause',
  volume: 'Volume',
  prev: 'Previous',
  next: 'Next',
  cast: 'Chromecast',
  fullscreen: 'Fullscreen',
  playlist: 'Playlist',
  hd: 'Quality',
  cc: 'Closed captions',
  audioTracks: 'Audio tracks', // need to convert camelCase to hyphens
  replay: 'Replay',
  buffer: 'Loading'
}
```

File: src/js/view/controlbar.js [hunk 2]
```diff
@@ -81,42 +89,47 @@ define([
 
         build : function() {
             var timeSlider = new TimeSlider(this._model, this._api),
-                drawer = new Drawer('jw-icon-more'),
+                drawer = new Drawer('jw-icon-more', this._aria['jw-icon-more']),
```
Comment (robwalch) [inline src/js/view/controlbar.js]: I didn't see 'more' in the list above.

File: src/js/view/controlbar.js [hunk 3]
```diff
@@ -81,42 +89,47 @@ define([
...
                 playlistTooltip,
                 volumeSlider,
                 volumeTooltip,
                 muteButton;
 
             // Create the playlistTooltip as long as visualplaylist from the config is not false
             if(this._model.get('visualplaylist') !== false) {
-                playlistTooltip = new Playlist('jw-icon-playlist');
+                playlistTooltip = new Playlist('jw-icon-playlist', this._aria['jw-icon-playlist']);
```
Comment (robwalch) [inline src/js/view/controlbar.js]: `.playlist`

**Comment by francoismassart** [comment]
Hey @robwalch about the `tabindex` usage, it should not cause any issue.
I added more explainations inside the [accessibility.md](https://github.com/francoismassart/jwplayer/blob/f237d0ce7a140319ce611d73036ba41e59458a65/accessibility.md)

`tabindex="0"` means "please allow this html node to be accessed via the tab key (even if it is a `<div>`), but don't change the order, just use the same order as the markup"

`tabindex="-1"` means "please ignore this html node when the user uses the tab key"

`tabindex="1"`, `tabindex="2"`, etc. (any other value than "0" or "-1") should never be used because it messes with the natural order defined by the markup and it'll quickly become nearly impossible to manage and can create unexpected tab jumps...

You can read more about this here:
[Don’t Use Tabindex Greater than 0](http://adrianroselli.com/2014/11/dont-use-tabindex-greater-than-0.html)

**As you will read inside the `accessibility.md`, there is still other issues to fix for improving the accessibility in JWPlayer but this would be already nice to have.**

Can you merge or provide additional feedback ?

Cheers

File: src/js/view/components/tooltip.js [hunk 4]
```diff
@@ -4,9 +4,25 @@ define([
...
-        'constructor' : function(name) {
+        'constructor' : function(name, ariaText, ariaShown) {
             this.el = document.createElement('div');
             this.el.className = 'jw-icon jw-icon-tooltip ' + name + ' jw-button-color jw-reset jw-hidden';
+            if (ariaText) {
+                this.el.setAttribute('tabindex', '0');
+                this.el.setAttribute('role', 'button');
+                this.el.setAttribute('aria-label', ariaText);
+            }
+            if (ariaShown === true) {
```
Comment (johnBartos) [inline src/js/view/components/tooltip.js]: more succinct to write as `if (ariaShown)`

**Comment by robwalch** [comment]
Thank you @francoismassart. My team needs to demo the new features to our product manager to get it our roadmap.

I'd like us to set the milestone of 7.5 for this, but it might have to wait until later. I'll keep you posted.

**Comment by francoismassart** [comment]
@robwalch, I can make a quick screencast demonstrating the before/after user experience if you want.
Just let me know :^)

**Comment by dannyfinks** [comment]
test this please

**Pull request merged**